### PR TITLE
1801/h/tech 1796 event timing callback

### PR DESCRIPTION
**Pull request by ColinDKelley**
@peterwilson and @caitlinrhd  Could you please review?

EventMachine part of https://ringrevenue.atlassian.net/browse/TECH-1796.

Build on 1.0-stable, which was 1.0.9.1 vs. 1.06.3 that we were using.

### CODE REVIEW

File: ext/cmain.cpp [hunk 1]
```diff
@@ -107,6 +107,17 @@ extern "C" const uintptr_t evma_install_oneshot_timer (int seconds)
...
 
 
+/*****************
+evma_get_real_time
+******************/
+
+extern "C" VALUE evma_get_real_time ()
+{
+	ensure_eventmachine("evma_get_real_time");
+	return INT2FIX(EventMachine->GetRealTime ());
```
Comment (ColinDKelley) [inline ext/cmain.cpp:117]: This is just EventMachine's tick counter so I can stamp the tick samples with it.

File: ext/em.cpp [hunk 2]
```diff
@@ -175,8 +175,8 @@ EventMachine_t::~EventMachine_t()
 	close (LoopBreakerWriter);
 
 	// Remove any file watch descriptors
-	while(!Files.empty()) {
-		map<int, Bindable_t*>::iterator f = Files.begin();
+	while (!Files.empty()) {
+		Files_t::iterator f = Files.begin();
```
Comment (ColinDKelley) [inline ext/em.cpp:179]: I changed to `for` and `while` loops and DRYed up the types with `typedef`.

File: ext/rubymain.cpp [hunk 3]
```diff
@@ -88,7 +91,7 @@ static inline VALUE ensure_conn(const uintptr_t signature)
 t_event_callback
 ****************/
 
-static inline void event_callback (struct em_event* e)
+static inline VALUE event_callback (struct em_event* e)
```
Comment (ColinDKelley) [inline ext/rubymain.cpp:95]: I changed the contract here to return the symbol of the method that was called so I could put that in the tick timing.

File: ext/rubymain.cpp [hunk 4]
```diff
@@ -1368,18 +1411,5 @@ extern "C" void Init_rubyeventmachine()
 
 	rb_define_method (EmConnection, "get_outbound_data_size", (VALUE(*)(...))conn_get_outbound_data_size, 0);
 	rb_define_method (EmConnection, "associate_callback_target", (VALUE(*)(...))conn_associate_callback_target, 1);
-
```
Comment (ColinDKelley) [inline ext/rubymain.cpp]: It made no sense to define these Ruby constants from C! I moved them out to Ruby and made a hash that maps them to their symbols.

File: lib/eventmachine.rb [hunk 5]
```diff
@@ -960,22 +982,22 @@ def self.connection_count
...
+  # the tick queue entries that were there when we started; new tick events scheduled by tick events
   # have to wait for the next pass through the reactor core.
   #
   # @private
   def self.run_deferred_callbacks
     until (@resultqueue ||= []).empty?
-      result,cback = @resultqueue.pop
-      cback.call result if cback
+      result, cback = @resultqueue.pop
+      cback.call(result)
```
Comment (ColinDKelley) [inline lib/eventmachine.rb:995]: I changed this contract to only put the callbacks on the queue when they're non-nil.

File: lib/eventmachine.rb [hunk 6]
```diff
@@ -1065,8 +1087,10 @@ def self.spawn_threadpool
             break # Ruby 2.0 may fail at Queue.pop
           end
           result = op.call
-          @resultqueue << [result, cback]
-          EventMachine.signal_loopbreak
+          if cback
```
Comment (ColinDKelley) [inline lib/eventmachine.rb:1090]: The corresponding contract change.

**Review by peterwilson — APPROVED** [review]
Can't say I know the inner workings of EM super well, but looks fine to me

File: ext/em.cpp [hunk 7]
```diff
@@ -505,13 +506,9 @@ void EventMachine_t::_DispatchHeartbeats()
 	// is changed out from underneath MyCurrentLoopTime.
 	const EventableDescriptor *head = NULL;
 
-	while (true) {
-		multimap<uint64_t,EventableDescriptor*>::iterator i = Heartbeats.begin();
```
Comment (caitlinrhd) [inline ext/em.cpp:509]: This used to run `Heartbeats.begin()` in every loop, and now it just runs it when entering the loop. Are there any negative implications of that change?

**Review by caitlinrhd — COMMENTED** [review]
Just one question.

File: ext/em.cpp (hunk 7 above)
Comment (ColinDKelley) [reply ext/em.cpp:509]: Good catch! It doesn't look like my code is equivalent. I will put it back as they had it.

**Review by caitlinrhd — APPROVED** [review]
👍

**Pull request closed without merging**